File: Frontend/worldflow/lib/Data/Models/post.dart
import 'comment.dart';

class Post {
  String id;
  String title;
  String ownerId;
  String ownerName;
  List<Comment> comments;
  List<String> likes;
  List<String> dislikes;
  DateTime createdAt;

  Post({
    required this.id,
    required this.title,
    required this.ownerId,
    required this.ownerName,
    required this.comments,
    required this.likes,
    required this.dislikes,
    required this.createdAt,
  });

  factory Post.fromJson(Map<String, dynamic> json) {
    // print('likes : ');
    // print(json['interactions']['likes']);

    List<Comment> comments = [];
    List<String> likes = [];
    List<String> dislikes = [];

    var tempComments = json['comments'] as List<dynamic>;
    comments = tempComments.map((e) => Comment.fromJson(e)).toList();

    var tempLikes = json['interactions']['likes'] as List<dynamic>;
    likes = tempLikes.map((e) => e.toString()).toList();

    var tempDislikes = json['interactions']['dislikes'] as List<dynamic>;
    dislikes = tempDislikes.map((e) => e.toString()).toList();

    return Post(
      id: json['_id'],
      title: json['title'],
      ownerId: json['ownerId'],
      ownerName: json['ownerName'],
      comments: comments,
      likes: likes,
      dislikes: dislikes,
      // convert string to DateTime
      createdAt: DateTime.parse(json['time']),
      // createdAt: json['time'],
    );
  }
}

File: Backend/LIB/Models/post.js
const mongoose = require('mongoose');

const PostSchema = mongoose.Schema({
    _id: mongoose.Schema.Types.ObjectId, //! ID is automatically created by MongoDB - ID OF POST
    title: {
        type: String,
        required: true,
        },  
    // content: { //! CONTENT OF POST
    //     type: String,
    //     minlength: 10,
    //     required: true,
    //     },  
    ownerId : { //! ID OF USER WHO MADE THE POST
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true,
        },
    ownerName : { //! NAME OF USER WHO MADE THE POST
        type: String,
        required: true,
        },
    time : { //! TIME OF POST CREATION
        type: Date,
        default: Date.now,
        },
    comments : [], //! COMMENTS OF POST
    interactions: { //! INTERACTIONS OF POST
        likes: { //! LIKES OF POST
            type: Array,
            default: [],
            },
        dislikes: { //! DISLIKES OF POST
            type: Array,
            default: [],
            },
        }
});

const Post = mongoose.model('Post', PostSchema);
module.exports = Post;